### DISCUSSION TITLE
More quickly select one of the URL suggestions

### DISCUSSION THREAD

**Discussion by AckslD** [qutebrowser/qutebrowser#8099] [Ideas]
I'm not sure something related to this has been discussed so though I'd open a discussion first. I tried to search a bit but wasn't really sure what terms to look for.

Consider the following, lets say I often visit some website `www.foo.bar` and then moving around in the website many different URLs are populate in the history, maybe something like:
```
www.foo.bar/tnfwaostra/stre
www.foo.bar/ttsrneaitsr/stratsrat
www.foo.bar/ttsrneaitsr/stratsrat/strstarstars
...
www.foo.bar
...
www.foo.bar/tsr
www.foo.bar/tfwffpwaaa
```
Now when I start qutebrowser and want to navigate to `www.foo.bar` I type `:open` and then start typing `foo`. In many cases I get something like the above as the suggestions and the one I want to go to, ie `www.foo.bar`, is somewhere in the middle. It's then hard to type more things to filter down further since it's the shortest match and all the other contain at least that URL.

I then have some options:

1. Use the mouse.
2. Press tab until I get to the URL I want.
3. Add a bookmark/quickmark to the URL.

Maybe 3. is the way to go but the same issue can occur in that case if I actually have many urls (eg gitlab repos) with very similar names.
Also I really don't want to do 1. :) and so 2. is what I often do but I'm wonder we can have something faster.

I'm thinking something similar to for example the [hop](https://github.com/phaazon/hop.nvim) plugin for `nvim` which has a `HopLine` command. For example if I have the buffer

![gscreenshot_2024-02-10-142247](https://github.com/qutebrowser/qutebrowser/assets/23341710/949d0d14-fe12-4df6-b748-29092ef51f74)

and do `HopLine`, each line gets associated with a character

![gscreenshot_2024-02-10-142301](https://github.com/qutebrowser/qutebrowser/assets/23341710/bc08f2ec-218a-4dfd-a110-0bf1400bfea1)

which I can press to quickly jump to that line.

I'm wondering if something similar could be done for URL suggestions (or any command suggestions) where I can press some key-command that takes me to this mode and I can go to any suggestion with a single key-press after that. I guess the key-command would have to be some `ctrl-...` since the user is at that point in command-mode. Of course the suggestions could also be identified by numbers in order but at least I'm faster at typing letters that numbers.

Firstly I'm just wondering if something similar has been considered and if this sounds good/reasonable. Then I'm happy to discuss more details and implementation and possibly help out with that.

**Comment by toofar** [comment]
I'm pretty sure that's not possible currently. Here is a feature request for it: https://github.com/qutebrowser/qutebrowser/issues/6883

> Press tab until I get to the URL I want.

Just in case you weren't aware, you can bind other keys to `completion-item-focus next` in command mode, in case pressing tab in particular is an issue! You can also bind something to eg `cmd-repeat 14 completion-item-focus next` to go down a few rows.

> It's then hard to type more things to filter down further since it's the shortest match and all the other contain at least that URL.

That resonates with me, I get that all the time and keep refusing to add them to bookmarks because it feels like too much commitment.

Another possible approach apart from hints would be to have some way to add custom commands that could operate on the completion by filtering it or selecting a specific item out of the list. Then you could bind something to filter to only domains, or select the shortest visible entry with the same domain as the selected one, or change the sorting. That's not possible currently either.

Not sure how hard it would be to add hinting to the completion. Probably involve both the completion widget and model. Might be a bit tricky, but at least be somewhat self contained, apart from the dozen dependant models!

**Reply by AckslD** [reply]
> Here is a feature request for it: https://github.com/qutebrowser/qutebrowser/issues/6883

Thanks for sharing that, I subscribed to it.

> Just in case you weren't aware, you can bind other keys to completion-item-focus next in command mode, in case pressing tab in particular is an issue! You can also bind something to eg cmd-repeat 14 completion-item-focus next to go down a few rows.

Thanks, yeah I think in my case tab is especially bad here. I have tab be a normal tab on press but alt on hold, so I can't in this case hold it for a bit to move down some. This is of course my own fault :)

> Another possible approach apart from hints would be to have some way to add custom commands that could operate on the completion by filtering it or selecting a specific item out of the list. Then you could bind something to filter to only domains, or select the shortest visible entry with the same domain as the selected one, or change the sorting. That's not possible currently either.

Yeah, sorting could potentially be useful as well, but I think hinting would be the best. Since in most cases I can look at the one I want but just not select it fast enough. If I see the one I want and then sort, I need to first find it again and then select it.

**Comment by mschilli87** [comment]
> 3. Add a bookmark/quickmark to the URL.
>
> Maybe 3. is the way to go but the same issue can occur in that case if I actually have many urls (eg gitlab repos) with very similar names.

I do that. Regarding gitlab urls, I don't type 'gitlab' but the name of the repo and then I get the corresponding bookmark. Since I never browse the list of all bookmarks like one might do in Firefox or so, having a lot of bookmarks does not have any drawbacks for me. So whenever I find myself in the situation you describe, I just navigate to the 'correct' entry *once* via option 2 and the bookmark it.

One other option would be a configuration option to sort the suggestions in a way that is mire likely to give you the correct one on top, *eg.* by length.

**Reply by AckslD** [reply]
> I do that. Regarding gitlab urls, I don't type 'gitlab' but the name of the repo and then I get the corresponding bookmark. Since I never browse the list of all bookmarks like one might do in Firefox or so, having a lot of bookmarks does not have any drawbacks for me. So whenever I find myself in the situation you describe, I just navigate to the 'correct' entry once via option 2 and the bookmark it.

Thanks, yeah I also don't type `gitlab` but in many cases I have some simlarly named repos `foo-core`, `foo-tools`, `bar-tools` etc. In this case I can distinguish them but I still think I would find it useful to quickly jump to one selection once I see it :)
